r/src/bpcells-cpp/matrixIterators/H5DenseMatrixWriter.h

```diff
@@ -0,0 +1,143 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#pragma once
+
+#include "../arrayIO/array_interfaces.h"
+#include "../arrayIO/bp128.h"
+#include "../arrayIO/hdf5.h"
+#include "../utils/filesystem_compat.h"
+#include "MatrixIterator.h"
+#include "OrderRows.h"
+
+namespace BPCells {
+
+template <typename T> class H5DenseMatrixWriter : public MatrixWriter<T> {
+  private:
+    HighFive::DataSet dataset;
+    HighFive::DataType datatype = HighFive::create_datatype<T>();
+    uint64_t max_capacity;
+    bool row_major;
+
+    static HighFive::DataSet createH5Matrix(
+        HighFive::File file,
+        std::string dataset_path,
+        uint64_t nrow,  // nrow of input matrix
+        uint64_t ncol,  // ncol of input matrix
+        bool row_major,
+        uint64_t chunk_size,
+        uint32_t gzip_level
+    ) {
+      HighFive::SilenceHDF5 s;
+      // Create a dataspace with initial shape and max shape
+      uint64_t nrow_h5 = nrow;
+      uint64_t ncol_h5 = ncol;
+      if (row_major) {
+          nrow_h5 = ncol;
+          ncol_h5 = nrow;
+      }
+      HighFive::DataSpace dataspace({nrow_h5, ncol_h5}, {HighFive::DataSpace::UNLIMITED, HighFive::DataSpace::UNLIMITED});
+
+      // Use chunking
+      HighFive::DataSetCreateProps props;
+      props.add(HighFive::Chunking(std::vector<hsize_t>{std::min(chunk_size, nrow_h5), std::min(chunk_size, ncol_h5)}));
+      if (gzip_level > 0) {
+          props.add(HighFive::Shuffle());
+          props.add(HighFive::Deflate(gzip_level));      
+      }
+
+      if (file.exist(dataset_path)) {
+          file.unlink(dataset_path);
+      }
+      // Create the dataset
+      return file.createDataSet<T>(dataset_path, dataspace, props);
+    }
+
+  public:
+    H5DenseMatrixWriter(
+        const HighFive::File &file,
+        std::string path,
+        uint64_t nrow,  // nrow of input matrix
+        uint64_t ncol,  // ncol of input matrix
+        bool row_major,
+        uint64_t max_capacity = 16384,
+        uint64_t chunk_size = 1024,
+        uint32_t gzip_level = 0
+    )
+        : dataset(createH5Matrix(file, path, nrow, ncol, row_major, chunk_size, gzip_level))
+        , max_capacity(max_capacity) 
+        , row_major(row_major) {}
+
+    void write(MatrixLoader<T> &mat_in, std::atomic<bool> *user_interrupt = NULL) override {
+
+        // Ensure that we write matrices sorted by row
+        OrderRows<T> mat((std::unique_ptr<MatrixLoader<T>>(&mat_in)));
+        // Don't delete our original matrix
+        mat.preserve_input_loader();
+
+        uint32_t col = 0;
+        uint64_t idx = 0; // Index of for col_ptr array
+        std::vector<T> zero_buf(mat_in.rows());
+        std::vector<T> val_buf = zero_buf;  // buffer for each column
+        while (mat.nextCol()) {
+            if (user_interrupt != NULL && *user_interrupt) return;
+            if (mat.currentCol() < col)
```

Comment:
I think `mat.currentCol() != col` wouldn't work because in the case of a matrix with a column of all zeros, it's possible to have `col < mat.currentCol()`. (That said, I'm not sure the case of a missing column is being handled correctly right now)